xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -525,18 +528,49 @@ where
                             published_in_epoch,
                             group_epoch
                         );
+                        let processing_error = if published_in_epoch < group_epoch {
+                            GroupMessageProcessingError::OldEpoch(
+                                published_in_epoch as u64,
+                                group_epoch as u64,
+                            )
+                        } else {
+                            GroupMessageProcessingError::FutureEpoch(
+                                published_in_epoch as u64,
+                                group_epoch as u64,
+                            )
+                        };
 
-                        return Err(Ok(IntentState::ToPublish));
+                        return Err(IntentValidationError {
+                            processing_error,
+                            next_intent_state: IntentState::ToPublish,
+                        });
                     }
 
-                    let staged_commit = if let Some(staged_commit) = &intent.staged_commit {
-                        match decode_staged_commit(staged_commit) {
-                            Err(err) => return Err(Err(err)),
-                            Ok(staged_commit) => staged_commit,
-                        }
-                    } else {
-                        return Err(Err(GroupMessageProcessingError::IntentMissingStagedCommit));
-                    };
+                    let staged_commit = intent
+                        .staged_commit
+                        .as_ref()
+                        .map_or(
+                            Err(GroupMessageProcessingError::IntentMissingStagedCommit),
+                            |staged_commit| decode_staged_commit(staged_commit),
+                        )
+                        .map_err(|err| {
+                            // If we can't retrieve the cached staged commit from the intent, we can't
+                            // apply it. It is indeterminate whether other members were able to apply it
+                            // or not - if they did apply it, then we are forked.
+                            tracing::error!(
+                                inbox_id = self.context.inbox_id(),
+                                installation_id = %self.context.installation_id(),
+                                group_id = hex::encode(&self.group_id),
+                                cursor = cursor,
+                                intent_id = intent.id,
+                                intent.kind = %intent.kind,
+                                "Error decoding staged commit for intent, now may be forked: {err:?}",
+                            );
+                            IntentValidationError {
+                                processing_error: err,
+                                next_intent_state: IntentState::Error,
+                            }
```

Comment:
This is the only functional change in this PR as far as I can tell. Previously, an error here would return an `Err(err)`, which will get bubbled up by the caller on L1245 of the old code.

Now, because the error is non-retriable, the caller will swallow it on L1279 of the new code and update the intent state to `Error`.